shell/server/admin/system-status-server.js

```diff
@@ -1,9 +1,61 @@
-const Crypto = Npm.require("crypto");
+import { Meteor } from "meteor/meteor";
+import { Router } from "meteor/iron:router";
+import { Random } from "meteor/random";
+import Crypto from "crypto";
+import Fs from "fs";
 
 const hashSessionId = function (sessionId) {
   return Crypto.createHash("sha256").update(sessionId).digest("base64");
 };
 
+// An in-memory map of random id -> timestamp, used to authorize requests to download the system
+// log.
+const SYSTEM_LOG_DOWNLOAD_TOKENS = {};
+
+Meteor.methods({
+  adminGetServerLogDownloadToken() {
+    const db = this.connection.sandstormDb;
+    if (!db.isAdminById(this.userId)) {
+      throw new Meteor.Error(403, "User must be admin to download system log.");
+    }
+
+    const token = Random.id();
+    SYSTEM_LOG_DOWNLOAD_TOKENS[token] = Date.now();
+    return token;
+  },
+});
+
+Router.map(function () {
+  this.route("adminDownloadServerLog", {
+    where: "server",
+    path: "/admin/status/server-log/:tokenId",
+    action() {
+      const token = this.params.tokenId;
+      const response = this.response;
+      const issueDate = SYSTEM_LOG_DOWNLOAD_TOKENS[token];
+      SYSTEM_LOG_DOWNLOAD_TOKENS[token] = undefined; // Clear the token from the in-memory map.
+      if (issueDate === undefined || issueDate + 60000 < Date.now()) {
+        // Require download to start within 60 seconds of the adminGetServerLogDownloadToken
+        // method being called.
+        response.writeHead(404, { "Content-Type": "text/plain" });
+        return response.end("Invalid server log download token.");
+      }
+
+      const logFilePath = SANDSTORM_LOGDIR + "/sandstorm.log";
+      // Lazily assume that this logfile won't be "too large" and we can just load it into memory
```

Comment:
Yeah, there's a couple catches like "what if the file is being extended as we stream it" but that should be easy enough.

I'd missed the `fd` option to [fs.createReadStream](https://nodejs.org/docs/latest-v0.10.x/api/fs.html#fs_fs_createreadstream_path_options) the first time I read through the docs, which made me fearful of TOCTOU mismatch.

I'm implement streaming now.